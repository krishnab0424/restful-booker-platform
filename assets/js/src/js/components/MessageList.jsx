import React from 'react';
import { API } from '../libs/Api.js';
import Message from './Message.jsx';

export default class MessageList extends React.Component {

    constructor() {
        super();

        this.state = {
            messageId : 0,
            messages : []
        }

        this.deleteMessage = this.deleteMessage.bind(this);
        this.refreshMessageList = this.refreshMessageList.bind(this);
        this.openMessage = this.openMessage.bind(this);
        this.closeMessage = this.closeMessage.bind(this);
    }

    componentDidMount() {
        this.refreshMessageList()
    }

    deleteMessage(id) {
        API.deleteMessage(id, this);

        this.props.setCount();
    }

    refreshMessageList(){
        API.getMessages(this);
    }

    openMessage(id){
        this.setState({messageId : id});
    }

    closeMessage(){
        this.refreshMessageList();
        this.setState({messageId : 0});

        this.props.setCount();
    }

    render() {
        return <div>
            {this.state.messageId > 0 && 
                <Message messageId={this.state.messageId} closeMessage={this.closeMessage} refreshMessageList={this.refreshMessageList} />
            }
            <div className="messages">
                <div className="row">
                    <div className="col-sm-2 rowHeader"><p>Name</p></div>
                    <div className="col-sm-9 rowHeader"><p>Subject</p></div>
                    <div className="col-sm-1"></div>
                </div>
                {this.state.messages.map((value, index) => {
                    return  <div className={"row detail read-" + value.read} id={"message" + index} key={index}>
                                <div className="col-sm-2" onClick={() => this.openMessage(value.id)}><p>{value.name}</p></div>
                                <div className="col-sm-9" onClick={() => this.openMessage(value.id)}><p>{value.subject}</p></div>
                                <div className="col-sm-1">
                                <span className="fa fa-remove roomDelete" onClick={() => this.deleteMessage(value.id)}></span>
                                </div>
                            </div>
                })}
            </div>
        </div>
    }

}