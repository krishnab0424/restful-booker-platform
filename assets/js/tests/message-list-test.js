import React from 'react';
import MessageList from '../src/js/components/MessageList.jsx';
import nock from 'nock';
import { waitForState } from 'enzyme-async-helpers';

beforeAll(() => {
    nock('http://localhost')
        .persist()
        .get('/message/')
        .reply(200, {
                messages : [
                    {
                        "name" : "Mark Winteringham",
                        "subject" : "Subject description here",
                        "read" : true
                    }, {
                        "name" : "James Dean",
                        "subject" : "Another description here",
                        "read" : false
                    }, {
                        "name" : "Janet Samson",
                        "subject" : "Lorem ipsum dolores est",
                        "read" : true
                    }
                ]
            }
        );
});

test('Renders the list of messages correctly', async () => {
    const messageComponent = shallow(<MessageList setCount={() => {}} />);

    await waitForState(messageComponent, state => state.messages.length === 3);

    expect(messageComponent).toMatchSnapshot();
});

test('Deletes message when selected to delete', (done) => {
    nock('http://localhost')
        .delete('/message/1')
        .reply(201, () => {
            done();
        });

    const messageComponent = shallow(<MessageList setCount={() => {}} />);
    messageComponent.instance().deleteMessage(1);
});

test('Clicking message shows message popup', async () => {
    const messageComponent = shallow(<MessageList setCount={() => {}} />);

    await waitForState(messageComponent, state => state.messages.length === 3);
    messageComponent.find("#message0").simulate('click');

    expect(messageComponent).toMatchSnapshot();
});
