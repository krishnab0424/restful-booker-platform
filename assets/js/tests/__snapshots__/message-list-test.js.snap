// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicking message shows message popup 1`] = `
<div>
  <div
    className="messages"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-2 rowHeader"
      >
        <p>
          Name
        </p>
      </div>
      <div
        className="col-sm-9 rowHeader"
      >
        <p>
          Subject
        </p>
      </div>
      <div
        className="col-sm-1"
      />
    </div>
    <div
      className="row detail read-true"
      id="message0"
      key="0"
    >
      <div
        className="col-sm-2"
        onClick={[Function]}
      >
        <p>
          Mark Winteringham
        </p>
      </div>
      <div
        className="col-sm-9"
        onClick={[Function]}
      >
        <p>
          Subject description here
        </p>
      </div>
      <div
        className="col-sm-1"
      >
        <span
          className="fa fa-remove roomDelete"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="row detail read-false"
      id="message1"
      key="1"
    >
      <div
        className="col-sm-2"
        onClick={[Function]}
      >
        <p>
          James Dean
        </p>
      </div>
      <div
        className="col-sm-9"
        onClick={[Function]}
      >
        <p>
          Another description here
        </p>
      </div>
      <div
        className="col-sm-1"
      >
        <span
          className="fa fa-remove roomDelete"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="row detail read-true"
      id="message2"
      key="2"
    >
      <div
        className="col-sm-2"
        onClick={[Function]}
      >
        <p>
          Janet Samson
        </p>
      </div>
      <div
        className="col-sm-9"
        onClick={[Function]}
      >
        <p>
          Lorem ipsum dolores est
        </p>
      </div>
      <div
        className="col-sm-1"
      >
        <span
          className="fa fa-remove roomDelete"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the list of messages correctly 1`] = `
<div>
  <div
    className="messages"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-2 rowHeader"
      >
        <p>
          Name
        </p>
      </div>
      <div
        className="col-sm-9 rowHeader"
      >
        <p>
          Subject
        </p>
      </div>
      <div
        className="col-sm-1"
      />
    </div>
    <div
      className="row detail read-true"
      id="message0"
      key="0"
    >
      <div
        className="col-sm-2"
        onClick={[Function]}
      >
        <p>
          Mark Winteringham
        </p>
      </div>
      <div
        className="col-sm-9"
        onClick={[Function]}
      >
        <p>
          Subject description here
        </p>
      </div>
      <div
        className="col-sm-1"
      >
        <span
          className="fa fa-remove roomDelete"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="row detail read-false"
      id="message1"
      key="1"
    >
      <div
        className="col-sm-2"
        onClick={[Function]}
      >
        <p>
          James Dean
        </p>
      </div>
      <div
        className="col-sm-9"
        onClick={[Function]}
      >
        <p>
          Another description here
        </p>
      </div>
      <div
        className="col-sm-1"
      >
        <span
          className="fa fa-remove roomDelete"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="row detail read-true"
      id="message2"
      key="2"
    >
      <div
        className="col-sm-2"
        onClick={[Function]}
      >
        <p>
          Janet Samson
        </p>
      </div>
      <div
        className="col-sm-9"
        onClick={[Function]}
      >
        <p>
          Lorem ipsum dolores est
        </p>
      </div>
      <div
        className="col-sm-1"
      >
        <span
          className="fa fa-remove roomDelete"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
